### PR TITLE
commit for FINERACT-65 (Implement ability to schedule & e-mail reports)

### PR DESCRIPTION
**Pull request by emmanuelnnaa**

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/domain/ReportRepositoryWrapper.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            throw new ReportNotFoundException(id);
+        }
+        
+        return report;
+    }
+    
+    /**
+     * @return {@link ReportRepository} object
+     */
+    public ReportRepository getReportRepository() {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/domain/ReportRepositoryWrapper.java:58]: I don't see any references to this method. Is this method required?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadReportingServiceImpl.java [hunk 2]
```diff
@@ -485,4 +487,10 @@ public ReportParameterData mapRow(final ResultSet rs, @SuppressWarnings("unused"
         }
     }
 
-}
\ No newline at end of file
+    @Override
+    public ByteArrayOutputStream generatePentahoReportAsOutputStream(String reportName, String outputTypeParam,
+            Map<String, String> queryParams, Locale locale, AppUser runReportAsUser, StringBuilder errorLog) {
+        // TODO Auto-generated method stub
+        return null;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadReportingServiceImpl.java:494]: What is the use of this method? Just return null?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadReportingService.java [hunk 3]
```diff
@@ -44,4 +47,7 @@
     ReportData retrieveReport(final Long id);
 
     Collection<String> getAllowedReportTypes();
+    
+    ByteArrayOutputStream generatePentahoReportAsOutputStream(String reportName, String outputTypeParam,
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadReportingService.java:51]: I don't see any reference to this method. Is it required as implementation returns just null

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java [hunk 4]
```diff
@@ -2736,4 +2736,27 @@ public CommandWrapperBuilder deleteSelfServiceBeneficiaryTPT(final Long benefici
...
\ No newline at end of file
+	public CommandWrapperBuilder createReportMailingJob(final String entityName) {
+        this.actionName = "CREATE";
+        this.entityName = entityName;
+        this.entityId = null;
+        this.href = "/reportmailingjobs";
+        return this;
+    }
+    
+    public CommandWrapperBuilder updateReportMailingJob(final String entityName, final Long entityId) {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java:2747]: Caller is passing permission name and entity id. But you are assigning permission name to entity name

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java [hunk 5]
```diff
@@ -2736,4 +2736,27 @@ public CommandWrapperBuilder deleteSelfServiceBeneficiaryTPT(final Long benefici
...
+        this.actionName = "UPDATE";
+        this.entityName = entityName;
+        this.entityId = entityId;
+        this.href = "/reportmailingjobs/" + entityId;
+        return this;
+    }
+    
+    public CommandWrapperBuilder deleteReportMailingJob(final String entityName, final Long entityId) {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java:2755]: Caller is passing permission name and entity id. But you are assigning permission name to entity name.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/api/ReportMailingJobApiResource.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        final ApiRequestJsonSerializationSettings settings = this.apiRequestParameterHelper.process(uriInfo.getQueryParameters());
+        final ReportMailingJobData ReportMailingJobDataOptions = this.reportMailingJobReadPlatformService.retrieveReportMailingJobEnumOptions();
+        
+        return this.reportMailingToApiJsonSerializer.serialize(settings, ReportMailingJobDataOptions, ReportMailingJobConstants.REPORT_MAILING_JOB_DATA_PARAMETERS);
+    }
+    
+    @GET
+    @Consumes({ MediaType.APPLICATION_JSON })
+    @Produces({ MediaType.APPLICATION_JSON })
+    public String retrieveAllReportMailingJobs(@Context final UriInfo uriInfo) {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/api/ReportMailingJobApiResource.java:149]: Suggestion: Better to give pagination support to retrieve all report mailing jobs.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/domain/ReportMailingJob.java [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+import org.joda.time.DateTime;
+import org.joda.time.LocalDate;
+import org.joda.time.LocalDateTime;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+import org.springframework.data.jpa.domain.AbstractPersistable;
+
+@Entity
+@Table(name = "m_report_mailing_job", uniqueConstraints = { @UniqueConstraint(columnNames = { "name" }, name = "unique_name") })
+public class ReportMailingJob extends AbstractPersistable<Long> {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/domain/ReportMailingJob.java:50]: Updated user and date are missing. Extend this entity from AbstractAuditableCustom

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/domain/ReportMailingJobConfigurationRepositoryWrapper.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            throw new ReportMailingJobConfigurationNotFoundException(name);
+        }
+        
+        return reportMailingJobConfiguration;
+    }
+    
+    /** 
+     * @return ReportMailingJobConfigurationRepository JPA repository object
+     **/
+    public ReportMailingJobConfigurationRepository getReportMailingJobConfigurationRepository() {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/domain/ReportMailingJobConfigurationRepositoryWrapper.java:53]: Is this method required?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/domain/ReportMailingJobRepositoryWrapper.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            throw new ReportMailingJobNotFoundException(id);
+        }
+        
+        return reportMailingJob;
+    }
+    
+    /** 
+     * @return ReportMailingJobRepository Jpa Repository object
+     **/
+    public ReportMailingJobRepository getReportMailingJobRepository() {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/domain/ReportMailingJobRepositoryWrapper.java:53]: Is this method required?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/service/ReportMailingJobWritePlatformServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+     * @param fileName
+     * @param byteArrayOutputStream
+     * @param errorLog
+     */
+    private void sendReportFileToEmailRecipients(final ReportMailingJob reportMailingJob, final String fileName, 
+            final ByteArrayOutputStream byteArrayOutputStream, final StringBuilder errorLog) {
+        final Set<String> emailRecipients = this.reportMailingJobValidator.validateEmailRecipients(reportMailingJob.getEmailRecipients());
+        
+        try {
+            final File file = new File(fileName);
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/service/ReportMailingJobWritePlatformServiceImpl.java:479]: Don't you need to close this file instance once mail is sent?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/domain/ReportMailingJobRepositoryWrapper.java (hunk 9 above)
Comment (emmanuelnnaa) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/domain/ReportMailingJobRepositoryWrapper.java:53]: @nazeer1100126 

Both methods are used in the "org.apache.fineract.infrastructure.reportmailingjob.service.ReportMailingJobWritePlatformServiceImpl" class.

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java (hunk 5 above)
Comment (emmanuelnnaa) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java:2755]: @nazeer1100126 

permission name = "REPORTMAILINGJOB" which is the same as the entity name.

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java (hunk 4 above)
Comment (emmanuelnnaa) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java:2747]: @nazeer1100126 

permission name = "REPORTMAILINGJOB" which is the same as the entity name.

**Comment by emmanuelnnaa** [comment]
@nazeer1100126 
Will create a new pull request.

**Pull request closed without merging**